doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -659,168 +734,476 @@ insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.
+      :ref:`Example<ltac2_match_vs_multimatch_ex>`.
+
+   :n:`@cpattern`
+      The syntax of :token:`cpattern` is
+      the same as that of :token:`term`\s, but it can contain pattern matching
+      metavariables in the form :n:`?@ident`.  :g:`_` can be used to match
+      irrelevant terms.
+
+      .. todo Example is broken :ref:`Example<ltac2_match_with_holes_ex>`.
+
+      .. todo Didn't understand the following 2 paragraphs well enough to revise
+         see https://github.com/coq/coq/pull/12103#discussion_r436297754 for a
+         possible example
+
+      When a metavariable in the form :n:`?id` occurs under binders,
+      say :n:`x__1, …, x__n` and the expression matches, the
+      metavariable is instantiated by a term which can then be used in any
+      context which also binds the variables :n:`x__1, …, x__n` with
+      same types. This provides with a primitive form of matching under
+      context which does not require manipulating a functional term.
+
+      There is also a special notation for second-order pattern matching: in an
+      applicative pattern of the form :n:`@?@ident @ident__1 … @ident__n`,
+      the variable :token:`ident` matches any complex expression with (possible)
+      dependencies in the variables :n:`@ident__i` and returns a functional term
+      of the form :n:`fun @ident__1 … @ident__n => @term`.
+
+   .. _match_term_context:
+
+   :n:`context {? @ident } [ @cpattern ]`
+      Matches any term with a subterm matching :token:`cpattern`. If there is a match
+      and :n:`@ident` is present, it is assigned the "matched
+      context", i.e. the initial term where the matched subterm is replaced by a
+      hole.  Note that `context`
+      (with very similar syntax) appearing after the `=>` is the :tacn:`context` tactic.
+
+      For :tacn:`match!` and :tacn:`multi_match!`, if the evaluation of the :token:`ltac2_expr`
+      fails, the next matching subterm is tried. If no further subterm matches, the next branch
+      is tried.  Matching subterms are considered from top to bottom and from left to
+      right (with respect to the raw printing obtained by setting the
+      :flag:`Printing All` flag).  :ref:`Example<ltac2_match_term_context_ex>`.
+
+   .. todo There's a more realistic example from @JasonGross here:
+      https://github.com/coq/coq/pull/12103#discussion_r432996954
+
+   :n:`@ltac2_expr`
+      The tactic to apply if the construct matches.  Metavariable values from the pattern
+      match are substituted
+      into :n:`@ltac2_expr` before it's applied.  Note that metavariables are not
+      prefixed with the question mark as they are in :token:`cpattern`.
+
+      If :token:`ltac2_expr` evaluates to a tactic, then it is
+      applied. If the tactic succeeds, the result of the match expression is
+      :n:`()`.  If :token:`ltac2_expr` does not evaluate
+      to a tactic, that value is the result of the match expression.
```

Comment:
What?  This paragraph makes no sense.  Was it taken directly from the Ltac1 chapter?  The semantics of Ltac1 and Ltac2 are quite different, and Ltac2 (thankfully) does not have two separate phases like Ltac1 does.

Reply:
> Was it taken directly from the Ltac1 chapter?

Pretty much.  As I said [here](url), "I rewrote the sections for `match!` and `match! goal` using the Ltac1 text as a starting point."  It's a real struggle guessing what Ltac2 is intended to do or how it's supposed to differ from Ltac1.  I hope you can tell me which aspects don't apply to Ltac2.

Reply:
From a bird's eye view: I believe it took a while to convey how Ltac1 has two separate evaluation phases, and how it's used to both run tactics and return values through a weird semantics.  This is confusing, AFAICT, to literally everyone who encounters it.  Ltac2 removes all of this in favor of the OCaml way of doing things; mutating the goal is considered a side-effect, and otherwise the semantics are effectively the same as OCaml's (e.g., side-effect-only functions return unit, functions can have both a return value and side effects, etc).  The one complication is that effects on the goal live inside the backtracking monad, while other side-effects do not, as seen by:
```coq
Require Import Ltac2.Ltac2.
Require Ltac2.Control Ltac2.Int Ltac2.Message.


Ltac2 foo () :=
  let v := { contents := 0 } in
  let incr () := ((v.(contents) := Int.add (v.(contents)) 1);
                 Message.print (Message.of_int (v.(contents)))) in
  Control.plus incr (fun _ => incr ()); Control.zero Not_found.

Goal True.
  try (foo ()). (* 1 2 *)
```

Ltac2 is basically about exposing both Ltac1 features and the lower-level OCaml tactic API, but doing away with the home-brewed strange semantics of Ltac1.

Reply:
Yeah.  This chapter will need a couple more rounds of revision before it's really good.  My main goal in this round is to get the syntax right.

I have pondered whether we could evolve Ltac1 semantics so they're not so strange.  For example, could we:
- add the `$var` construct to Ltac1 (optional, not required)
- add an option that, at run time, reports constructs that lack the appropriate `$var`
- add an option to require use of `$var` and drop the related strange semantic
and change the defaults for the options over a release or two?

It seems to me there's a lot of Ltac1 code out there, that converting is non-trivial and it's not clear (at least to me) when Ltac2 will be really ready for prime time and well-documented.  Incremental improvements to Ltac1 along these lines seems like it would be less disruptive for users.